common/optimiser/Optimiser.java

```diff
@@ -45,30 +53,49 @@ public class Optimiser {
     private MPSolver solver;
     private MPSolverParameters parameters;
     private Status status;
-    private boolean hasSolver;
-    private double objectiveValue;
+    private boolean isConstraintsUpToDate;
+    private Double objectiveValue;
 
     public Optimiser() {
         variables = new ArrayList<>();
         constraints = new HashSet<>();
         objectiveCoefficients = new HashMap<>();
         status = Status.NOT_SOLVED;
-        hasSolver = false;
+        objectiveValue = null;
+        isConstraintsUpToDate = false;
     }
 
     public synchronized Status optimise(long timeLimitMillis) {
-        assert hasSolver;
         if (isOptimal()) return status;
+        assert solutionSatisfiesModel();
+        maySetSolver();
```

Comment:
We got rid of an API: `restartFromSolution()`. From the point of view of users of `Optimiser`, this is an implementation-specific method that we had to call to update the solver whenever the `Constraint` coefficients had changed, or we created the model for the first time. However, this can all be transparently be handled.

Now, whenever we call `optimise()`, we already know if the solver needs to be initialised from scratch versus the coefficients on constraints need to be updated only. This leads to a cleaner UX of the Optimiser class.